Job with concatenator fails using run first N records

I have a job which contains a Plain search/replace, a Concatenator and an output write, which runs successfully when executed normally, but fails when run using the run first N records option. It results in a stacktrace which in itself doesn't explain very much:

```
13:48:16ERROR: An error occurred in the analysis job!

org.datacleaner.job.ComponentValidationException
	at org.datacleaner.descriptors.ValidateMethodDescriptorImpl.convertThrownException(ValidateMethodDescriptorImpl.java:42)
	at org.datacleaner.descriptors.AbstractMethodDescriptor.invoke(AbstractMethodDescriptor.java:83)
	at org.datacleaner.descriptors.ValidateMethodDescriptorImpl.validate(ValidateMethodDescriptorImpl.java:37)
	at org.datacleaner.lifecycle.LifeCycleHelper.validate(LifeCycleHelper.java:160)
	at org.datacleaner.job.tasks.InitializeTask.executeInternal(InitializeTask.java:71)
	at org.datacleaner.job.tasks.InitializeTask.execute(InitializeTask.java:55)
	at org.datacleaner.job.concurrent.TaskRunnable.run(TaskRunnable.java:61)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
```

--- a/engine/core/src/main/java/org/datacleaner/job/builder/AbstractComponentBuilder.java
+++ b/engine/core/src/main/java/org/datacleaner/job/builder/AbstractComponentBuilder.java
@@ -445,19 +445,19 @@ public abstract class AbstractComponentBuilder<D extends ComponentDescriptor<E>,
 
                                 if (dependentValue != null) {
                                     // First build a list containing value and references tuples.
-                                    final Map<Integer, Object> originalMappings = new HashMap<>();
+                                    final Map<String, Object> originalMappings = new HashMap<>();
 
                                     final List<Object> synchronizedDependents = new ArrayList<>();
 
                                     if (currentValue.getClass().isArray()) {
                                         for (int i = 0; i < Array.getLength(currentValue); i++) {
-                                            originalMappings.put(Array.get(currentValue, i).hashCode(), Array.get(
+                                            originalMappings.put(getKey(Array.get(currentValue, i)), Array.get(
                                                     dependentValue, i));
                                         }
 
                                         for (int i = 0; i < Array.getLength(newValue); i++) {
-                                            synchronizedDependents.add(originalMappings.get(Array.get(newValue, i)
-                                                    .hashCode()));
+                                            synchronizedDependents.add(originalMappings.get(getKey(Array.get(newValue,
+                                                    i))));
                                         }
 
                                         dependentProperty.setValue(_configurableBean, getArray(dependentProperty
@@ -470,6 +470,18 @@ public abstract class AbstractComponentBuilder<D extends ComponentDescriptor<E>,
                                 }
                             });
         }
+    }
+
+    private static String getKey(final Object object) {
+        if (object instanceof InputColumn<?>) {
+            final InputColumn<?> inputColumn = (InputColumn<?>) object;
+
+            if (inputColumn.isVirtualColumn()) {
+                return inputColumn.getName();
+            }
+        }
+
+        return String.valueOf(object.hashCode());
     }
 
     private static <E> E[] getArray(Class<E> clazz, List<?> baseList) {

--- a/engine/core/src/test/java/org/datacleaner/job/builder/InputColumnLinkingTest.java
+++ b/engine/core/src/test/java/org/datacleaner/job/builder/InputColumnLinkingTest.java
@@ -24,6 +24,7 @@ import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
 
 import org.apache.metamodel.schema.MutableColumn;
+import org.apache.metamodel.schema.MutableTable;
 import org.datacleaner.api.InputColumn;
 import org.datacleaner.api.MappedProperty;
 import org.datacleaner.components.convert.ConvertToDateTransformer;
@@ -43,7 +44,11 @@ public class InputColumnLinkingTest {
     @Before
     public void setup() {
         _jobBuilder = new AnalysisJobBuilder(new DataCleanerConfigurationImpl());
-        _jobBuilder.addSourceColumn(new MutableColumn(SOURCE_COLUMN_NAME));
+
+        final MutableColumn column = new MutableColumn(SOURCE_COLUMN_NAME);
+        column.setTable(new MutableTable("table"));
+
+        _jobBuilder.addSourceColumn(column);
 
         _dateTransformer = _jobBuilder.addTransformer(ConvertToDateTransformer.class);
         _dateTransformer.addInputColumn(_jobBuilder.getSourceColumnByName(SOURCE_COLUMN_NAME));
@@ -182,5 +187,33 @@ public class InputColumnLinkingTest {
         assertEquals(0, analyzer.getInputColumns().size());
         assertEquals(0, analyzer.getComponentInstance()._columns.length);
         assertEquals(0, analyzer.getComponentInstance()._columnNames.length);
+    }
+
+    /**
+     * Tests if all mapped properties are copied correctly when creating a copy of a job and the generated ids
+     * for the input column object identifiers are not in sync with the ones in the source job.
+     */
+    @Test
+    public void testCopyJob() {
+        // Add and remove at transformer to make sure the id's for job copies are out of sync with the source job id's.
+        _jobBuilder.removeTransformer(_jobBuilder.addTransformer(ConvertToStringTransformer.class));
+
+        final TransformerComponentBuilder<ConvertToStringTransformer> stringTransformer = _jobBuilder.addTransformer(
+                ConvertToStringTransformer.class);
+        stringTransformer.addInputColumn(_dateTransformer.getOutputColumns().get(0));
+
+        final AnalyzerComponentBuilder<MockAnalyzer> analyzer = _jobBuilder.addAnalyzer(MockAnalyzer.class);
+        final ConfiguredPropertyDescriptor columnsProperty = getPropertyDescriptor(analyzer, "Columns");
+
+        analyzer.addInputColumn(stringTransformer.getOutputColumns().get(0), columnsProperty);
+        analyzer.setConfiguredProperty("Column names", new String[] { "first" });
+
+        final AnalysisJobBuilder jobBuilderCopy = new AnalysisJobBuilder(_jobBuilder.getConfiguration(), _jobBuilder
+                .toAnalysisJob(false));
+
+        assertNotNull(((MockAnalyzer) jobBuilderCopy.getAnalyzerComponentBuilders().get(0)
+                .getComponentInstance())._columnNames[0]);
+
+        jobBuilderCopy.close();
     }
 }

